controller/supportbundle_controller.go

```diff
@@ -0,0 +1,456 @@
+	}
+}
+
+func (sbc *SupportBundleController) Create(sb *longhorn.SupportBundle, image string) error {
+
+	log := getLoggerForSupportBundle(sbc.logger, sb)
+
+	deployName := sbc.getManagerName(sb)
+	log.Debugf("creating deployment %s with image %s", deployName, image)
+
+	pullPolicy := sbc.getImagePullPolicy()
+
+	deployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      deployName,
+			Namespace: sb.Namespace,
+			Labels: map[string]string{
+				"app":                       types.SupportBundleManager,
+				types.SupportBundleLabelKey: sb.Name,
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					Name:       sb.Name,
+					Kind:       sb.Kind,
+					UID:        sb.UID,
+					APIVersion: sb.APIVersion,
+				},
+			},
+		},
+		Spec: appsv1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{"app": types.SupportBundleManager},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app":                       types.SupportBundleManager,
+						types.SupportBundleLabelKey: sb.Name,
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:            "manager",
+							Image:           image,
+							Args:            []string{"/usr/bin/support-bundle-utils", "manager"},
+							ImagePullPolicy: pullPolicy,
+							Env: []corev1.EnvVar{
+								{
+									Name:  "LONGHORN_NAMESPACE",
+									Value: sb.Namespace,
+								},
+								{
+									Name:  "LONGHORN_VERSION",
+									Value: manager.FriendlyVersion(),
+								},
+								{
+									Name:  "LONGHORN_SUPPORT_BUNDLE_NAME",
+									Value: sb.Name,
+								},
+								{
+									Name:  "LONGHORN_SUPPORT_BUNDLE_DEBUG",
+									Value: "true",
+								},
+								{
+									Name: "LONGHORN_SUPPORT_BUNDLE_MANAGER_POD_IP",
+									ValueFrom: &corev1.EnvVarSource{
+										FieldRef: &corev1.ObjectFieldSelector{
+											FieldPath: "status.podIP",
+										},
+									},
+								},
+								{
+									Name:  "LONGHORN_SUPPORT_BUNDLE_IMAGE",
+									Value: image,
+								},
+								{
+									Name:  "LONGHORN_SUPPORT_BUNDLE_IMAGE_PULL_POLICY",
+									Value: string(pullPolicy),
+								},
+							},
+							Ports: []corev1.ContainerPort{
+								{
+									ContainerPort: 8080,
+								},
+							},
+						},
+					},
+					ServiceAccountName: sbc.ServiceAccount,
+				},
+			},
+		},
+	}
+
+	_, err := sbc.ds.CreateDeployment(deployment)
+	return err
+}
+
+func (sbc *SupportBundleController) setError(sb *longhorn.SupportBundle, reason string) (*longhorn.SupportBundle, error) {
+	log := getLoggerForSupportBundle(sbc.logger, sb)
+	log.Errorf(reason)
+
+	toUpdate := sb.DeepCopy()
+	sbc.SupportBundleInitialized.False(toUpdate)
+	sbc.SupportBundleInitialized.Message(toUpdate, reason)
+
+	toUpdate.Status.State = types.SupportBundleStateError
+	return sbc.ds.UpdateSupportBundle(toUpdate)
```

Comment:
Don't call `UpdateSupportBundle` more than once in the controller. Follow the pattern in https://github.com/longhorn/longhorn-manager/blob/master/controller/volume_controller.go to only update on exiting the reconciliation loop.

Reply:
@yasker  updated.